File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py


def solution(begin, target, words):
    length = len(words[0])
    words.insert(0, begin)
    queue = [0]
    visited = [False for n in range(len(words))]
    visited[0] = True
    
    depths = [0 for n in range(len(words))]
    
    while queue:
        current = queue.pop(0)
        for next in range(len(words)):
            if visited[next]:
                continue
            diff = 0
            for i in range(length):
                if words[next][i] != words[current][i]:
                    diff += 1
            if diff == 1:
                depths[next] = depths[current] + 1
                if words[next] == target:
                    return depths[next]
                visited[next] = True
                queue.append(next)
    return 0
        

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    
    private boolean canChange(String a, String b) {
        char[] arrA = a.toCharArray();
        char[] arrB = b.toCharArray();
        int diff = 0;
        for(int i = 0; i < a.length(); i++) {
            if (arrA[i] != arrB[i]) diff++;
        }
        return diff == 1;
    }
    
    public int solution(String begin, String target, String[] words) {
        Queue<String> queue = new ArrayDeque<>();
        queue.add(begin);
        
        Map<String, Integer> visited = new HashMap<>();
        visited.put(begin, 0);
        
        while(!queue.isEmpty()) {
            String current = queue.poll();
            
            if(current.equals(target)) {
                return visited.get(target);
            }
            
            for(String word: words) {
                if(!visited.containsKey(word) && canChange(current, word)) {
                    queue.add(word);
                    visited.put(word, visited.get(current) + 1);
                }
            }
        }
        
        return 0;
    }
}